docs/CEK.md

```diff
@@ -0,0 +1,200 @@
+s , {_ A}                  <| abs α M    |-> s ; ρ |> M [ α / A ]*
+s , wrap A B _             <| V          |-> s <| wrap A B V
+s , unwrap _               <| wrap A B V |-> s <| V
+s , builtin b As Vs _ [] ρ <| V          |->
+ | bn computes on As (Vs ++ [V]) to V'      = s <| V'
+ | bn computes on As (VS ++ [V]) to error A = <> A
+s , builtin b As Vs _ (M ∷ Ms) ρ <| V    |->
+  s , builtin b As (Vs ++ [V]) _ Ms ρ ; ρ |> M
+```
+
+One can discharge a value back to a term. Note that this cannot change
+the head symbol of a value though, so a lambda value is never going to
+turn into a constructor etc. If we only want to check if the answer is
+true/false then it's not necessary to discharge, we can just look at
+the value.
+
+This version is, I think, the same as the version of CEK described in
+the 1986 Tech Report "Control Operators, the SECD-machine, and the
+λ-calculus" by Felleisen and Friedman, page 5, table 1,
+https://legacy.cs.indiana.edu/ftp/techreports/TR197.pdf. Their machine
+looks a bit different as they have 3-tuples of a term or nothing, an
+environment and a continuation for the state. Our `s ; ρ |> M` is
+written `(M , ρ , s)` and our `s <| V` is written `(%, 0, s ret V)`
+where % signifies there is no term in focus and 0 is the empty
+environment.
+
+## 2. Closure version
+
+This is the version described in the spec and also the version used in
+the implementation. I have also formalised this.
+
+Values are simpler with just terms under binders:
+```
+V := con cn | abs ɑ M | lam x M | iwrap A B V
+```
+Closures:
+```
+C := (V , ρ)
+```
+Environments contain closures:
+```
+ρ := [] | ρ[x |-> C]
+```
+Frames: note the contain closures instead of just values
+```
+f := [_ (M,ρ)]
+   | [C _]
+   | {_ A}
+   | wrap A B _
+   | unwrap _
+   | builtin b As Cs _ Ms ρ
+```
+
+Stacks of frames
+```
+s := .
+   | s , f
+```
+
+States contain environments in both the term and value configuration.
+```
+σ := s ; ρ |> M
+   | s ; ρ <| V
+   | <> A
+   | [] C
+```
+
+The machine. Term configuration:
+```
+s ; ρ |> x                      |-> s ; ρ' <| V where (V , ρ' ) = ρ[ x ]
+s ; ρ |> lam x L                |-> s ; ρ <| lam x L
+s ; ρ |> [L M]                  |-> s , [_ (M,ρ)]  ; ρ |> L
+s ; ρ |> abs α L                |-> s ; ρ <| abs α L
+s ; ρ |> {L A}                  |-> s , {_ A} ; ρ |> L
+s ; ρ |> wrap A B L             |-> s , wrap A B _ ; ρ |> L
+s ; ρ |> unwrap L               |-> s , unwrap _ ; ρ |> L
+s ; ρ |> con c                  |-> s ; ρ <| con c
+s ; ρ |> builtin b As           |-> s ; ρ' |> M
+  where bn computes on As to (M,ρ')
+s ; ρ |> builtin b As (M :: Ms) |-> s , builtin b As [] _ Ms ρ ; ρ |> M
+s ; ρ |> error A                |-> <> A
+```
+Value configuration:
+```
+.                                ; ρ <| V          |-> [] (V,ρ)
+s , [_ (M,ρ)]                    ; ρ <| V          |-> s , [(V,ρ) _] ; ρ' |> M
+s , [(lam x (M,ρ')) _]           ; ρ <| V          |-> s ; ρ [ x |-> V ] |> M
```

Comment:
Shouldn't this put a closure in the environment?

Reply:
Yes. thanks.